python/ETL/transfer.py

```diff
@@ -0,0 +1,160 @@
+    # result_data = src_cur.fetchall()
+
+    # # Insert or update trees in the target database
+    # tree_columns = [desc[0] for i, desc in enumerate(src_cur.description) if i < 47]
+    # for row in result_data:
+    #     tree_data = row[:47]
+    #     insert_or_update("trees", tree_columns, [tree_data],  dest_cur, dest_conn, action=action)
+
+    # # Insert species in the target database if not exists
+    # species_columns = [desc[0] for i, desc in enumerate(src_cur.description) if i >= 47]
+    # for row in result_data:
+    #     species_data = row[47:]
+    #     if species_data and species_data[0] is not None:  # Check if species data exists for the tree
+    #         insert_or_update("tree_species", species_columns, [species_data],  dest_cur, dest_conn, action=action)
+
+    ## update without join operation ---- compare efficiency with join operation
+    # update trees assoicated with each planter
+    for planter in planters_data:
+        planter_id = planter[0]  
+        # Fetch trees associated with the current planter
+        src_cur.execute("SELECT * FROM trees WHERE planter_id = %s", (planter_id,))
+        trees_data = src_cur.fetchall()
+        tree_columns = [desc[0] for desc in src_cur.description]
+        insert_or_update("trees", tree_columns, trees_data, dest_cur, dest_conn, action=action)
+
+        for tree in trees_data:
+            species_id = tree[29] #index of species_id = 29
+
+            src_cur.execute("SELECT * FROM tree_species WHERE id = %s", (species_id,))
+            species_data = src_cur.fetchone()
+            
+            if species_data:
+                species_columns = [desc[0] for desc in src_cur.description]
+                insert_or_update("tree_species", species_columns, [species_data],  dest_cur, dest_conn, action=action)
+
+    # #update trees with organizations directly -- but data missing?
+    # src_cur.execute("SELECT * FROM trees WHERE planting_organization_id = %s", (org_id,))
+    # trees_data = src_cur.fetchall()
+    # trees_columns = [desc[0] for desc in src_cur.description]
+    # insert_or_update("trees", trees_columns, trees_data, dest_cur, dest_conn)
+
+    src_cur.close()
+    dest_cur.close()
+    src_conn.close()
+    dest_conn.close()
+
+    return 
+
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Transfer data from source to target database.")
+    parser.add_argument("-t","--target_db", default=None, help="URI for the target PostgreSQL database.")
+    parser.add_argument("-s","--source_db", default=DATABASE_URL, help="URI for the source PostgreSQL database.")
+    parser.add_argument("-o","--org_id", required=True, type=int, help="ID of the target organization.")
+    parser.add_argument("-a","--action", default = False,type=bool, help="Whether to update the database.")
+    args = parser.parse_args()
+    transfer_data(target=args.target_db, source=args.source_db,organization_id=args.org_id, action=args.action)
```

Comment:
Can you change this to the pattern as here?
https://github.com/Greenstand/treetracker-airflow-dags/blob/360580afffd51021661c3e88a78f8bae1462ed72/lib/capture_export.py
https://github.com/Greenstand/treetracker-airflow-dags/blob/360580afffd51021661c3e88a78f8bae1462ed72/lib/capture_export_test.py

Here is a simple guide for this test pattern:
https://github.com/Greenstand/treetracker-airflow-dags#option-1-develop-without-installing-airflow

By doing this, we can:
* Make this function platform-agnostic, so it is easy to be reused by: cli, faas, airflow
* Easier to be tested, unit test lib used in the sample code is a good tool for testing and developing, you can easily check your code even hot-refresh your code once you change sometime in your sourcecode.

Reply:
Sure. I'll add the test file later.